### DISCUSSION TITLE
Would it be possible to change the Teensy for something cheaper?

### DISCUSSION THREAD

**Discussion by careck** [FundamentalFrequency/.github#15] [General]
As I understand, the Teensy is only grabbing keypresses and turning them into MIDI signals which are then sent to the rPi via USB.

Currently it's hard to get a hold of the Teensy 4.1 board and where it is available it's extremely expensive. At least here in Australia it goes for about $65 - $75 AUD.  In comparison, a Raspberry Pi 4B with 4Gb costs about $95 AUD.

So my question is, could a different MCU be used for this job? Or even a Teensy 4.0 which is a bit cheaper?

**Comment by stonepreston** [comment]
If you want you are welcome to try and use something different. I chose the Teensy 4.1 because that’s what PJRC recommends for new projects. 

You would have to update the PCB to use whatever footprint you choose. I can’t guarantee the firmware would work exactly as written for anything but the 4.1, so that might need updating as well. But if you stick to something within the teensy or even the broader arduino family you’ll probably be ok. 

I would ensure you pick something supported by [Control Surface](https://github.com/tttapa/Control-Surface) as it makes the firmware extremely simple. I know sending midi signals sounds pretty simple, but the logic behind reading the switch matrix and all the MIDI related stuff would be quite a bit to implement yourself, so stick with something supported by that (or a similar library)

You will also need to ensure whatever you pick has enough inputs. The current switch matrix is routed to be as easily understandable as possible at a glance (ie what looks like the rows and columns are the actual rows and columns), and is not optimized for using the minimum number of inputs, so if you really needed to you could reduce the number of required inputs some.

**Comment by benpocalypse** [comment]
I haven't done this in a long time (probably 10-ish years) but there are libraries that exist for a number of microcontrollers that let them act as USB-HID devices without too much trouble. The Teensy itself is rather costly considering it's basically just acting as a custom USB keyboard, so if anyone is up for a challenge, the could look at something like the V-USB library, or perhaps a project like this one:
http://blakesmith.me/2019/01/16/making-my-own-usb-keyboard-from-scratch.html

It could allow for a different base PCB for the LMN-3 that would accommodate a different footprint where the Teensy goes. I don't currently have the time to look into this, but it's certainly do-able.

**Reply by stonepreston** [reply]
So the reason I went with the Teensy is mainly how easy it is for anyone to write the firmware for it, and the fact it is through hole solderable to the board. This makes it very attractive from a DIY standpoint.

That being said I can definitely envision a version of the LMN down the line that uses a SMD microcontroller like an atmega32, and runs QMK firmware (QMK has SOME support for MIDI, last I checked it wasnt super extensive) or some other library. It would certainly be cheaper and clear up some space on the board. However, currently I think the priority of "DIYness" is important in these early stages, even if that means its more expensive.

**Reply by benpocalypse** [reply]
Absolutely agreed. The Teensy is by far the easiest option. I just know that I was lucky enough to have one lying around from before all the chip shortages, as it looks like they're in fairly short supply now.
Definitely wasn't meant to be a criticism, just an option for folks that want to get creative.

**Reply by stonepreston** [reply]
yeah the shortages are indeed a problem :/ maybe soon it will end

**Reply by benpocalypse** [reply]
Also, I owe you an apology. Until I had mine up and running the words "sends MIDI data to the LMN-3-DAW" hadn't sunk in. I had assumed since the Teensy was connected to traditional mechanical keyboard switches that it was simply acting as a USB keyboard. Disregard everything I previously said then :/

**Reply by stonepreston** [reply]
no its alright, its pretty similar. The library I mentioned earlier QMK is actually a library used mainly for writing USB keyboard firmware. It happens to have some MIDI functionality as well, so its a good fit for something like this. Although the MIDI functionality does not seem anywhere near the level the ControlSurface library I currently use provides, but it is something.

**Comment by DominusFulguris** [comment]
After analysing the current schematic, I think it is possible to reduce the pin count of the circuit sufficient to put the Raspberry Pi Pico forward as an alternative to the Teensy 4.1.

Current pin count is 30 pins (28 digital, 2 analog).  The design needs to be reduced to 26 pins for the Pico.

This can be done.

1. COL_0 is only being used by a single switch on ROW_4. Relocating this switch to a vacant node on ROW_3 an deleting COL_0 saves one pin.
2. The 25 transport control switches and encoder select switches are spread across ROW_0, ROW_1, ROW_2 with 14 columns (now 13 after COL_0 is removed).  If one row was eliminated (leaving 2 rows by 13 columns), the 25 key functions can then be remapped over the 26 available switch nodes.  Eliminating a row (nominally ROW_2) saves one pin.
3. As there are now four rows in total for the matrix, these could be further multiplexed using one additional chip (a 1-of-4 decoder such as 74HC139 series) to allow saving of two pins.
4. The Joystick SEL button could be remapped from a dedicated pin to a spare node on one of the two keyswitch matrices - presumably the last node on the transport matrix would make the most sense. Saves one pine.

These changes reduce the pin count in total to 25 and inside the pin budget available with the RP Pico. Item 4 above then becomes optional as there is still one pin available.

RP Pico is listed as supported by ControlSurface so firmware porting would hopefully be straight forward.

[Edit: remove reference to "dual 1-of-2 decoder"]

**Reply by careck** [reply]
This sounds really cool if it can be pulled off (not that I would be able to redesign the circuit board or even change the firmware). The pico costs only a 1/10 of the Teensy and would really bring the cost down considerably!

**Reply by stonepreston** [reply]
I think this would be a great move. Control surface does indeed say it supports the pico so yeah I dont think we'd run into many issues there. My only request would be ensure the multiplexer is through hole

**Reply by Sonosus** [reply]
I reckon an SPI multiplexer chip might be the way to go. Chips such as the MCP23S18 are cheap, readily available, and through hole. I second the Pico idea - it means people ordering a Pi 4 will probably be able to buy a Pico at the same shop.

**Reply by stonepreston** [reply]
The only thing that might be tricky is using the multiplexer with control surfaces button matrix, but it might work out of the box. Not sure

**Reply by DominusFulguris** [reply]
Looks like control surface does not support mux/demux in the button matrix code.  It does support such things in other areas.

Hacking in a change to do what is needed is easy, only made slightly more complex because there are two overlapping button matrixes.

Making the change in a non-hacky way (i.e. suitable for a pull request back into the repo) will require much more work.

Support could be done for an SPI I/O expander.  There is support in Control Surface for simple serial-to-parallel chips such as the '595 (but not in connection with button matrix).  The MCP23S18 is a great I/O expander chip (I have used it in other projects), but needs to be configured via registers before use which adds a lot of complexity to the problem.

**Reply by stonepreston** [reply]
we could roll our own matrix reading code, but it is really nice to let control surface handle everything. I wonder if there's any more optimization we can do to the matrix layout to squeeze a few extra pins out of it. Its insane how cheap the pico is compared to similar boards.

**Reply by DominusFulguris** [reply]
Switching to serial-to-parallel converters saves a lot of pins.

I have a design based on a PIC16F1459 (low pin count USB chip) which is supporting 16 LED-illuminated buttons.  It uses one 16-channel PWM LED controller and two 8-channel input mux chips in series.  Requires just 4 pins in total.  Want more I/O? Daisy chain in more chips - no additional pins required.  There is even PIC library code for doing USB MIDI, so code is easy.  Limitation is that it would be a bit too far out of main stream for an open source project - need a dedicated programmer, etc.

But even without going serial, we can easily add a 1-of-16 input mux on the columns which would drop you from 13 pins to just 5 (4 for channel select, 1 for data) to scan the columns.  The biggest problem is limiting yourself to through hole chips - which I rarely use these days.  The bulk of the cool chips are SMD.

**Reply by DominusFulguris** [reply]
1-of-16 input muxes are hard to come by and expensive.

So, if we use a pair of 74HC138 demux's to give 1-of-16 output selects and invert the matrix scanning (reverse all the diodes so columns are selects and rows are inputs to the MCU), I can get the keyboard scan happening with 8 pins: 4 output pins to the 74HC138's for columns, and 4 row inputs. Logically the switch matrix can be up to 16x4.

**Reply by DominusFulguris** [reply]
> we could roll our own matrix reading code, but it is really nice to let control surface handle everything. I wonder if there's any more optimization we can do to the matrix layout to squeeze a few extra pins out of it. Its insane how cheap the pico is compared to similar boards.

What would you want to use the extra pins for?

- MIDI (traditional) out? In?
- LEDs? (WS2812's are easy with this using single pin)
- I2S out? Could put an I2S DAC/amplifier on it
- Additional buttons or controller interfaces?
- Another display? I2C/SPI should be easy.

**Comment by Sonosus** [comment]
I'm planning on using an ATMEGA328P chip (along with some multiplexers) to replace the Teensy. I'll use a small serial-USB chip and Hairless MIDI to connect to the Raspberry PI. Now I come to think of it actually, it's probably possible to connect the serial interfaces of the atmega and the Pi together through a logic level shifterr to eliminate that complicated setup.
